File: windows/Whisper.Core/UpdateChecker.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;

namespace Whisper.Core;

/// <summary>
/// Checks GitHub Releases for a newer tagged version than the running app.
/// Mirrors the mac app's UpdateChecker: pure metadata lookup (including the
/// release's asset list, so <see cref="Updater"/> can pick out the .msi to
/// silently install) plus a fallback html_url for the "open the release page"
/// manual path. Historically this was check-only with no assets parsed --
/// that reasoning (no code-signed update channel) is the same one the mac
/// app outgrew: silently replacing an unsigned app with another copy of the
/// same unsigned app from the same publisher isn't a new trust boundary.
/// </summary>
public static class UpdateChecker
{
    public const string Repo = "Ofunrein/whisper";

    public sealed record Asset(string Name, string DownloadUrl);

    public sealed record Release(string TagName, string HtmlUrl, IReadOnlyList<Asset> Assets);

    public static async Task<Release?> CheckForUpdateAsync(string currentVersion, CancellationToken ct = default)
    {
        using var client = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };
        client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/vnd.github+json"));
        client.DefaultRequestHeaders.UserAgent.ParseAdd("Whisper-App");

        try
        {
            using var response = await client.GetAsync($"https://api.github.com/repos/{Repo}/releases/latest", ct);
            if (!response.IsSuccessStatusCode) return null;

            using var stream = await response.Content.ReadAsStreamAsync(ct);
            using var doc = await JsonDocument.ParseAsync(stream, cancellationToken: ct);
            var tagName = doc.RootElement.GetProperty("tag_name").GetString();
            var htmlUrl = doc.RootElement.GetProperty("html_url").GetString();
            if (string.IsNullOrEmpty(tagName) || string.IsNullOrEmpty(htmlUrl)) return null;

            var assets = new List<Asset>();
            if (doc.RootElement.TryGetProperty("assets", out var assetsEl) && assetsEl.ValueKind == JsonValueKind.Array)
            {
                foreach (var assetEl in assetsEl.EnumerateArray())
                {
                    var name = assetEl.TryGetProperty("name", out var nameEl) ? nameEl.GetString() : null;
                    var url = assetEl.TryGetProperty("browser_download_url", out var urlEl) ? urlEl.GetString() : null;
                    if (!string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(url))
                        assets.Add(new Asset(name, url));
                }
            }

            var latest = tagName.StartsWith('v') ? tagName[1..] : tagName;
            return IsNewer(latest, currentVersion) ? new Release(tagName, htmlUrl, assets) : null;
        }
        catch
        {
            return null;
        }
    }

    private static bool IsNewer(string a, string b)
    {
        var av = ParseVersion(a);
        var bv = ParseVersion(b);
        for (var i = 0; i < Math.Max(av.Length, bv.Length); i++)
        {
            var x = i < av.Length ? av[i] : 0;
            var y = i < bv.Length ? bv[i] : 0;
            if (x != y) return x > y;
        }
        return false;
    }

    private static int[] ParseVersion(string v) =>
        v.Split('.').Select(p => int.TryParse(p, out var n) ? n : 0).ToArray();
}

File: Sources/Whisper/Update/UpdateChecker.swift
import Foundation

/// Checks GitHub Releases for a newer tagged version than the running app.
///
/// Historically this was check-only: it just opened the release page in a
/// browser for the user to download manually, because the app isn't
/// notarized and a silently-replaced binary would hit the same Gatekeeper
/// quarantine wall again on next launch. `Updater` (see Updater.swift) now
/// handles that: it downloads the release DMG, mounts it, copies
/// Whisper.app to /Applications, and strips the quarantine attribute from
/// the freshly-copied app before relaunching -- which is safe because it's
/// a self-update of the same publisher's app the user already approved
/// once, not an arbitrary Gatekeeper bypass. The "open release page in
/// browser" behavior is kept as a manual fallback.
enum UpdateChecker {
    struct Asset: Decodable {
        let name: String
        let downloadURL: String

        enum CodingKeys: String, CodingKey {
            case name
            case downloadURL = "browser_download_url"
        }
    }

    struct Release: Decodable {
        let tagName: String
        let htmlURL: String
        let assets: [Asset]

        enum CodingKeys: String, CodingKey {
            case tagName = "tag_name"
            case htmlURL = "html_url"
            case assets
        }
    }

    static let repo = "Ofunrein/whisper"

    static var currentVersion: String {
        Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.0.0"
    }

    /// Returns the newer release if one exists, else nil.
    static func checkForUpdate() async -> Release? {
        guard let url = URL(string: "https://api.github.com/repos/\(repo)/releases/latest") else { return nil }
        var request = URLRequest(url: url)
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.timeoutInterval = 10

        guard let (data, response) = try? await URLSession.shared.data(for: request),
              let http = response as? HTTPURLResponse, http.statusCode == 200,
              let release = try? JSONDecoder().decode(Release.self, from: data) else {
            return nil
        }

        let latest = release.tagName.hasPrefix("v") ? String(release.tagName.dropFirst()) : release.tagName
        return isNewer(latest, than: currentVersion) ? release : nil
    }

    private static func isNewer(_ a: String, than b: String) -> Bool {
        let av = a.split(separator: ".").compactMap { Int($0) }
        let bv = b.split(separator: ".").compactMap { Int($0) }
        for i in 0..<max(av.count, bv.count) {
            let x = i < av.count ? av[i] : 0
            let y = i < bv.count ? bv[i] : 0
            if x != y { return x > y }
        }
        return false
    }
}
